Google.Api.Gax.Grpc/Rest/HttpRuleTranscoder.cs

```diff
@@ -9,89 +9,140 @@
+            .FirstOrDefault(result => result is not null);
+
+    /// <summary>
+    /// A transcoder for a single rule, ignoring additional bindings.
+    /// </summary>
+    private class SingleRuleTranscoder
+    {
+        private static readonly FieldType[] TypesIneligibleForQueryStringEncoding = { FieldType.Group, FieldType.Message, FieldType.Bytes };
+        private static readonly HttpMethod s_patchMethod = new HttpMethod("PATCH");
+
+        private readonly HttpMethod _httpMethod;
+        private readonly Func<IMessage, string> _bodyTranscoder;
+        private readonly HttpRulePathPattern _pathPattern;
+        private readonly List<FieldDescriptor> _queryParameterFields;
+
+        internal SingleRuleTranscoder(string methodName, MessageDescriptor requestMessage, HttpRule rule)
         {
-            // TODO: The body shouldn't contain any fields that are present in the URI.
-            "*" => (true, null, new Func<IMessage, string>(protoRequest => protoRequest.ToString())),
-            "" => (false, null, new Func<IMessage, string>(protoRequest => null)),
-            string name when requestMessageDescriptor.FindFieldByName(name) is FieldDescriptor field => (false, name, new Func<IMessage, string>(protoRequest => field.Accessor.GetValue(protoRequest).ToString())),
-            _ => throw new ArgumentException($"Method {methodName} has a body parameter {rule.Body} in the 'google.api.http' annotation which is not a field in {requestMessageDescriptor.Name}")
-        };
-
-        _queryParameterFields = new List<FieldDescriptor>();
-        if (!allFieldsInBody)
+            (string pattern, _httpMethod) = rule.PatternCase switch
+            {
+                HttpRule.PatternOneofCase.Get => (rule.Get, HttpMethod.Get),
+                HttpRule.PatternOneofCase.Post => (rule.Post, HttpMethod.Post),
+                HttpRule.PatternOneofCase.Patch => (rule.Patch, s_patchMethod),
+                HttpRule.PatternOneofCase.Put => (rule.Put, HttpMethod.Put),
+                HttpRule.PatternOneofCase.Delete => (rule.Delete, HttpMethod.Delete),
+                HttpRule.PatternOneofCase.Custom => (rule.Custom.Path, new HttpMethod(rule.Custom.Kind)),
+                _ => throw new ArgumentException("HTTP rule has no pattern")
+            };
+
+            _pathPattern = HttpRulePathPattern.Parse(pattern, requestMessage);
+
+            (bool allFieldsInBody, string bodyName, _bodyTranscoder) = rule.Body switch
+            {
+                // TODO: The body shouldn't contain any fields that are present in the URI.
+                "*" => (true, null, new Func<IMessage, string>(protoRequest => protoRequest.ToString())),
+                "" => (false, null, new Func<IMessage, string>(protoRequest => null)),
+                string name when requestMessage.FindFieldByName(name) is FieldDescriptor field => (false, name, new Func<IMessage, string>(protoRequest => field.Accessor.GetValue(protoRequest).ToString())),
```

Comment:
Can this ever be null?

`field.Accessor.GetValue(protoRequest)?.ToString() ?? ""`

Reply:
Or should that be a reason for failure? It's not clear on the AIP.